### ISSUE TITLE
Reduce computation of mineral concentration layer

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#935]
**Is your feature request related to a problem? Please describe.**
- Currently, in Mars Navigator, by checking the Mineral in the Map Option button, this mineral layer is made visible
- By default, the mineral layer is **always** visible/enabled in Mission Tool's Nav tab. There is no option to turn it off.

**Describe the solution you desire to see**
- It's possible that most of the time, player is interested in seeing the location of one mineral in nearby vicinity.
- Allow player to view one ore more mineral(s) at a time. 
- This more flexible feature should be available not just in Mars Navigator, but also in Mission Tool's Nav tab.

**Describe alternatives you've considered**
- May draw individual types of mineral scircles, square or triangular with border color, instead of solid colors, which take more computation.

**Describe the rationale**
- It was found in [#942](https://github.com/mars-sim/mars-sim/issues/942) that `MineralMapLayer` is very slow to render. Turning it off speed up the surface map rendering significantly.

**Screenshots**
![image](https://github.com/mars-sim/mars-sim/assets/1168584/46e5a82e-ca6b-4a0f-ad7d-80eee2472383)

**Specifications  (please complete the following information):**
 - pre 3.6.0 build 8254

**Additional context**
- Find ways to optimize displayLayer() method below.
```java
	/**
	 * Displays the layer on the map image.
	 * 
	 * @param mapCenter the location of the center of the map.
	 * @param baseMap   the type of map.
	 * @param g         graphics context of the map display.
	 */
	public void displayLayer(Coordinates mapCenter, Map baseMap, Graphics g) {
		String mapType = baseMap.getType().getId();
		if (!mapCenter.equals(mapCenterCache) || !mapType.equals(mapTypeCache) || updateLayer) {
			mapCenterCache = mapCenter;
			mapTypeCache = mapType;
			updateLayer = false;

			// Clear map concentration array.
			Arrays.fill(mineralConcentrationArray, 0);

			double centerX = Map.HALF_MAP_BOX;
			double centerY = Map.HALF_MAP_BOX;

			double rho = baseMap.getScale();

			java.util.Map<String, Color> mineralColors = getMineralColors();
			
			updateMineralsDisplayed();

			for (int x = 0; x < Map.MAP_VIS_WIDTH; x += 2) {
				for (int y = 0; y < Map.MAP_VIS_HEIGHT; y += 2) {
					Coordinates location = mapCenter.convertRectToSpherical(x - centerX, y - centerY, rho);
					java.util.Map<String, Double> mineralConcentrations = mineralMap
							.getAllMineralConcentrations(location);
					if (mineralConcentrations.isEmpty()) {
						continue;
					}
					Iterator<String> i = mineralConcentrations.keySet().iterator();
					while (i.hasNext()) {
						String mineralType = i.next();
						if (isMineralDisplayed(mineralType)) {
							double concentration = mineralConcentrations.get(mineralType);
							if (concentration <= 0) {
								continue;
							}
							Color baseColor = mineralColors.get(mineralType);
							int index = x + (y * Map.MAP_VIS_WIDTH);
							addColorToMineralConcentrationArray(index, baseColor, concentration);
							addColorToMineralConcentrationArray((index + 1), baseColor, concentration);
							if (y < Map.MAP_VIS_HEIGHT - 1) {
								int indexNextLine = x + ((y + 1) * Map.MAP_VIS_WIDTH);
								addColorToMineralConcentrationArray(indexNextLine, baseColor, concentration);
								addColorToMineralConcentrationArray((indexNextLine + 1), baseColor,
										concentration);
							}
						}
					}
				}
			}

			// Create mineral concentration image for map
			mineralConcentrationMap = displayComponent.createImage(new MemoryImageSource(Map.MAP_VIS_WIDTH,
					Map.MAP_VIS_HEIGHT, mineralConcentrationArray, 0, Map.MAP_VIS_WIDTH));

			MediaTracker mt = new MediaTracker(displayComponent);
			mt.addImage(mineralConcentrationMap, 0);
			try {
				mt.waitForID(0);
			} catch (InterruptedException e) {
				logger.log(Level.SEVERE, "MineralMapLayer interrupted: " + e);
				// Restore interrupted state
			    Thread.currentThread().interrupt();
			}
		}

		// Draw the mineral concentration image
		g.drawImage(mineralConcentrationMap, 0, 0, displayComponent);
	}
```

**Comment by Urwumpe** [comment]
Can we maybe also use a noise generator for producing more realistic mineral concentrations?

**Comment by mokun** [comment]
> Can we maybe also use a noise generator for producing more realistic mineral concentrations?

Sure. I haven't looked up how mineral should be typically distributed. But can you show me how a noise generator may work different from our methods in RandomUtil that use some implementation of pseudo random noise generator (PRNG)?

**Comment by mokun** [comment]
It turns out that the Mineral button already allows player to turn on and off each type of the 10 minerals.
 
![image](https://github.com/mars-sim/mars-sim/assets/1168584/80890c04-5cce-4b8d-92b4-7c6f0ce40f16)

But it's very difficult to view if a mineral is chosen or not because the depress state of being selected is not obvious under our Look & Feel setting. 

Will need to change it to a check mark or something to make it easy to view.

**Comment by bdowrick** [comment]
I'm currently working on a solution to speed up Mineral Map display (and some QoL improvements for performance for other map layers). 
I should be able to have a pull through in the next few days to reduce time to generate the frame by ~80%

The problem of performance is mostly limited to the below code:
```java
Coordinates location = mapCenter.convertRectToSpherical(x - centerX, y - centerY, rho);
java.util.Map<String, Double> mineralConcentrations = mineralMap.getAllMineralConcentrations(location);
```

It calls expensive calculations to get the Coordinate, and then iterates through *all* minerals to check the distance to these. I'm currently exploring different methods to speed up spatial calcs (RTree, Spatial Hash Grid, Quad Tree, etc...)

**Comment by Urwumpe** [comment]
> > Can we maybe also use a noise generator for producing more realistic mineral concentrations?
> 
> Sure. I haven't looked up how minerals should be typically distributed. But can you show me how a noise generator may work different from our methods in RandomUtil that use some implementation of pseudo random noise generator (PRNG)?

Take a look for example at Perlin noise or that library: [JNoise](https://github.com/Articdive/JNoise)

**Comment by mokun** [comment]
> It calls expensive calculations to get the Coordinate, and then iterates through all minerals to check the distance to these. 

You're right. 

Perhaps JoCL can help rewriting a better alternative of convertRectToSpherical()

Also, in my machine, getAllMineralConcentrations deals with a map of the size of 405001 elements. 

Even though it's expensive and very slow, somehow VM must be limiting or regulating in such a way that in my laptop the CPU util doesn't go above 25% or else it would affect other applications I'm running.

 > to speed up spatial calcs (RTree, Spatial Hash Grid, Quad Tree, etc...)

Sure. we haven't looked into those yet. @Urwumpe did mentione RTree and quad tree once or twice.

Thanks for spearheading into resolving that.

**Comment by mokun** [comment]
@bdowrick 

How far are you in completing your performance optimization ? 

I've done removing the Mars Globe, which is the left panel of Mars NavigatorWindow.

Also, I've removed a bunch of stuff from CannedMapData and just lightly modified IntegerMapData.

But before I commit, do you want me to wait for your pull and merge first, since yours is more crucial and difficult than mine?

Just let me know.

**Comment by mokun** [comment]
@bdowrick 

Good that we'll move forward with consolidating some class into a new submodule, as you've laid out and expounded in #947

Meanwhile, I'm removing MineralConcentration class that serves little purpose and cleaning up RandomMineralMap class.

**Comment by mokun** [comment]
I've finished consolidating some methods. 

Also switch the mineral concentration from double to integer.

However, the fact is that even after getting rid of using MineralConcentration class and substituting it with a map to store the mineral concentration, it doesn't help speeding up.

![image](https://github.com/mars-sim/mars-sim/assets/1168584/d36d1dfd-c503-4e57-9f07-cb1a97b85762)

We'll need to have a create a better way to draw that fuzzy blob around each mineral spot and their surrounding. It's too slow.

Anyway, I'll commit and we'll go from there to find ways to break that down further.

**Comment by mokun** [comment]
@bdowrick 

To speed up the rendering, I found the best is simply disable the mineral layer while we are dragging our mouse across the map.

The moment we release the mouse button, it will render the mineral layer once.

**Comment by mokun** [comment]
Another approach I'd like to try is to reuse the portion of the old array that was rendered previously and copy the color values onto the new array in the new mineral layer. 

It's not successful yet because I didn't get the index mapping (from the old mineralArray to the new mineralArray) right.

But may be you fix it.

**Comment by mokun** [comment]
@bdowrick 

Shall we move this issue to 3.8.0 ?

**Comment by bevans2000** [comment]
Closed as part of #1462
